apps/avatax/src/modules/webhooks/validate-order-confirmed-payload.ts

```diff
@@ -0,0 +1,23 @@
+import { TaxBadPayloadError } from "../taxes/tax-error";
+import { err, ok } from "neverthrow";
+import { OrderConfirmedPayload } from "./order-confirmed-payload";
+import { OrderStatus } from "../../../generated/graphql";
+
+export function verifyOrderConfirmedPayload(payload: OrderConfirmedPayload) {
+  if (!payload.order) {
+    return err(new TaxBadPayloadError("Insufficient order data"));
+  }
+
+  const hasBilling = !!payload.order?.billingAddress;
+  const hasAddress = !!payload.order?.shippingAddress;
+
+  if (!hasBilling && !hasAddress) {
+    return err(new TaxBadPayloadError("Order has no billing or shipping address"));
+  }
+
+  if (payload.order.status === OrderStatus.Fulfilled) {
+    return err(new TaxBadPayloadError("Skipping fulfilled order to prevent duplication"));
+  }
+
+  return ok(payload);
```

Comment:
to make it type safe, you can add function like "verifyAddressExist" that will return `Adress` object and you can set in typescript that it exstis.